compiler/rustc_mir_dataflow/src/framework/engine.rs

```diff
@@ -197,21 +217,32 @@ where
             tcx,
             apply_trans_for_block,
             pass_name,
+            blocks,
             ..
         } = self;
 
         let mut dirty_queue: WorkQueue<BasicBlock> = WorkQueue::with_none(body.basic_blocks.len());
+        // could be tracked in `entry_sets`
+        let mut visited = BitSet::new_empty(body.basic_blocks.len());
 
         if A::Direction::IS_FORWARD {
-            for (bb, _) in traversal::reverse_postorder(body) {
-                dirty_queue.insert(bb);
+            match blocks {
+                Blocks::All => {
+                    for (bb, _) in traversal::reverse_postorder(body) {
+                        dirty_queue.insert(bb);
+                    }
+                }
+                Blocks::Reachable => {
+                    dirty_queue.insert(mir::START_BLOCK);
+                }
             }
```

Comment:
The existing implementation already considers only reachable blocks, since `reverse_postorder` and `postorder` return only reachable blocks. Furthermore, is an error to query data flow state for unreachable blocks (there is a debug only assertion).

Though, it looks like the initial seeding of the work queue uses a different definition of reachability that the one in fixpoint computation. In particular, the fixpoint iteration ignores some unwind edges as configured by `dead_unwinds` method. Maybe this is the cause of the issue you described in https://github.com/rust-lang/rust/pull/104488#issuecomment-1365802291?

Reply:
If this indeed the issue, I would suggest removing dead unwind edges directly in MIR and getting rid of `Engine::dead_unwinds` functionality.

Reply:
> Though, it looks like the initial seeding of the work queue uses a different definition of reachability that the one in fixpoint computation. In particular, the fixpoint iteration ignores some unwind edges as configured by dead_unwinds method

Yep, that's exactly what I meant

> The existing implementation already considers only reachable blocks, since reverse_postorder and postorder return only reachable blocks. Furthermore, is an error to query data flow state for unreachable blocks

Oh, interesting. I remember seeing an error somewhere when I made `Blocks::Reachable` the default but can't seem to replicate it now.

> If this indeed the issue, I would suggest removing dead unwind edges directly in MIR and getting rid of Engine::dead_unwinds functionality.

That would be nice but I'm trying to keep modifications to a minimum here to maintain the focus and to me it seems a bit more work. I think a reasonably clean and easy change would be indeed to just put `mir::START_BLOCK` inside the queue at startup, what do you think?